Options docs, defaults and valid values

wathen opened the pull request:
#### Description of Work

Fixes #583 

Removes `default_options.ini` and moves the documentation into the read the docs and the default values to the options class. Also, adds `valid` values for certain options.

Docs are [here](https://fitbenchmarking.readthedocs.io/en/583_options_docs/)

#### Testing Instructions

1. Check new docs are complete with respect to the old default options file
2. Check docs for adding new options 
3. Check that there are not references to `default_options.ini` 

Function: Does the change do what it's supposed to?

Tests: Does it pass? Is there adequate coverage for new code?

Style: Is the coding style consistent? Is anything overly confusing?

Documentation: Is there a suitable change to documentation for this change?

wathen commented on docs/source/extending/options_extend.rst:
```diff
@@ -0,0 +1,25 @@
+
+1. Add new option to the corresponding ``DEFAULTS`` dictionary.
+2. **OPTIONAL** add accepted option values to the ``VALID`` dictionary.
+   options class.
```
Removed

wathen commented on docs/source/users/jacobian.rst:
```diff
@@ -4,10 +4,12 @@
 ################
 
-Currently FitBenchmarking uses SciPy's `numerical finite difference <https://github.com/scipy/scipy/blob/912c54cd6473225c902377df410258839511b863/scipy/optimize/_numdiff.py#L198>`_ approximation to the Jacobian. The finite difference schemes implemented are:
+Currently FitBenchmarking uses SciPy's `numerical finite difference <https://github.com/scipy/scipy/blob/912c54cd6473225c902377df410258839511b863/scipy/optimize/_numdiff.py#L198>`_ approximation to the Jacobian and uses analytical Jacobians for the CUTEst data set.
```
As discussed, we will merge this into master and then backport as is. After the backport we will remove analytical from the release

wathen commented on docs/source/users/options/fitting_option.rst:
```diff
@@ -0,0 +1,122 @@
+
+This sets the options type of Jacobian used. Current Jacobian methods are:
+
+* ``anaylytic`` - This uses the analytic Jacobian extracted from the fitting problem.
```
See above

wathen commented on docs/source/users/options/fitting_option.rst:
```diff
@@ -0,0 +1,122 @@
+
+**Note:** Currently analytic Jacobians are available are only available for
+problems that use the cutest parser, which are currently in the directories
+'SIF' and 'DIAMOND_SIF'.# default 'SciPyFD'
```
Fixed

wathen commented on docs/source/extending/options_extend.rst:
```diff
@@ -0,0 +1,25 @@
+
+    - Add user defined tests to class called ``User{Section}OptionTests``. These
+      should check that the user added option is set correctly and that if
+      there are accepted options an ``OptionsError`` is raised.
```
Done

wathen commented on fitbenchmarking/utils/options.py:
```diff
@@ -26,66 +93,88 @@ def __init__(self, file_name=None):
+            if not value_check:
+                self.error_message.append(
+                    "Value set in the options, {0}, is not in the VALID "
+                    "list {1}".format(value, self.VALID[section][option]))
```
Done

tyronerees commented on docs/source/users/options/fitting_option.rst:
```diff
@@ -0,0 +1,122 @@
+
+This sets the options type of Jacobian used. Current Jacobian methods are:
+
+* ``anaylytic`` - This uses the analytic Jacobian extracted from the fitting problem.
```
```suggestion
* ``analytic`` - This uses the analytic Jacobian extracted from the fitting problem.
```

wathen commented:
> I spotted one small typo, otherwise looks good.
> 
> Shall I write an issue for checking the keys int he ini file are supported?

I fixed that actually and added tests to check that the 4 sections are set correctly as well as the keys in the ini file
